torch.complex and torch.polar

wjgan7 opened the pull request:
For https://github.com/pytorch/pytorch/issues/35312 and https://github.com/pytorch/pytorch/issues/38458#issuecomment-636066256.

wjgan7 committed:
Initial complex code

wjgan7 commented:
@anjali411 @ezyang Hi, I saw the update about having the polar keyword, but was wondering, is something like what I wrote along the right track to implement `torch.complex`?

wjgan7 commented on aten/src/ATen/native/ComplexHelper.cpp:
```diff
@@ -0,0 +1,65 @@
+  } else {
+    Tensor real = at::mul(input1, at::cos(input2));
+    Tensor imag = at::mul(input1, at::sin(input2));
+    at::add_out(result, real, at::mul(imag, i));
```
I see, I'll look into that. I'm guessing my complex (non-polar version) should also be direct, and that complex_polar shouldn't rely on complex either?

anjali411 replied:
yeah that's correct

wjgan7 committed:
torch.complex and torch.complex_polar via kernels

wjgan7 commented:
Hi @anjali411 @mruberry, made some updates. A couple question/comments,

- Right now `Vec256` isn't supported, but I'm not sure if it's possible since a real and imaginary `Vec256<float>` will lead to two `Vec256<c10:complex<float>>`'s.
- How could I go about supporting the function for CUDA?
- For the derivatives, I saw a lot of entries in `derivatives.yaml` don't use a `_backward` kernel. I'm thinking about just adding an entry like
```
- name: complex(Tensor real, Tensor imag) -> Tensor
  real: grad
  imag: grad * c10::complex<float>{0.0, 1.0}

- name: complex_polar(Tensor abs, Tensor angle) -> Tensor
  abs: grad * (angle.cos() + c10::complex<float>{0.0, 1.0} * angle.sin())
  angle: grad * abs * (- angle.sin() + c10::complex<float>{0.0, 1.0} * angle.cos())
```
However, would it be better to have specific `complex_backward` and `complex_polar_backward` methods that do a more optimized version of this? (essentially same way how `complex` and `complex_polar` kernels are implemented right now)

mruberry commented:
> Hi @anjali411 @mruberry, made some updates. A couple question/comments,
> 
> * Right now `Vec256` isn't supported, but I'm not sure if it's possible since a real and imaginary `Vec256<float>` will lead to two `Vec256<c10::complex<float>>`'s.
> * How could I go about supporting the function for CUDA?
> * For the derivatives, I saw a lot of entries in `derivatives.yaml` don't use a `_backward` kernel. I'm thinking about just adding an entry like
> 
> ```
> - name: complex(Tensor real, Tensor imag) -> Tensor
>   real: grad
>   imag: grad * c10::complex<float>{0.0, 1.0}
> 
> - name: complex_polar(Tensor abs, Tensor angle) -> Tensor
>   abs: grad * (angle.cos() + c10::complex<float>{0.0, 1.0} * angle.sin())
>   angle: grad * abs * (- angle.sin() + c10::complex<float>{0.0, 1.0} * angle.cos())
> ```
> 
> However, would it be better to have specific `complex_backward` and `complex_polar_backward` methods that do a more optimized version of this? (essentially same way how `complex` and `complex_polar` kernels are implemented right now)

- It's OK to not support Vec256. 
- You're halfway there. Check out the binary CUDA kernels that use TensorIterator in aten/src/ATen/native/cuda. For example https://github.com/pytorch/pytorch/blob/fcc9a1e664a3a06ad2de5a3fe5eb94c46dd6bbd8/aten/src/ATen/native/cuda/BinaryAddSubKernel.cu#L11.
- I think it's OK to not write a custom backward kernel for now. Let's focus on correctness.

mruberry commented:
Also see: https://github.com/pytorch/pytorch/pull/40246/files, which might give you some ideas for manipulating the complex types.

wjgan7 commented:
@mruberry @anjali411 sorry for just getting back to this. I think backward won't work altogether since it's fitting a complex gradient into a float input tensor. This was making me wonder how this function should work. Should it be able to accept complex inputs as well (and just use their real components)? It also made me wonder if this function would really be useful, since my guess is once x and y are both complex, the speed advantage over x + 1j * y wouldn't be that great.

anjali411 commented:
> @mruberry @anjali411 sorry for just getting back to this. I think backward won't work altogether since it's fitting a complex gradient into a float input tensor. This was making me wonder how this function should work. Should it be able to accept complex inputs as well (and just use their real components)? It also made me wonder if this function would really be useful, since my guess is once x and y are both complex, the speed advantage over x + 1j * y wouldn't be that great.

@wjgan7 I think backward definition should be similar to that of `view_as_complex` ( cc. @albanD ).

The function should only be able to expect floating point tensors (cc. @mruberry). For cpu, you can define a function in this file: `aten/src/ATen/native/cpu/zmath.h` which can be passed to the `cpu_kernel`. The function would accept two floating points a and b and return a `c10::complex<T>(a, b)`. check out this PR as an example: https://github.com/pytorch/pytorch/pull/39955

albanD commented:
For the backward formula, I think we can just wait on the formula Anjali is going to put in the doc and apply it here. That will contain all the real to complex/complex to real changes that need to be applied.

wjgan7 commented:
@anjali411 I believe I already have the non-Vec256 implementation in `native/cpu/ComplexHelperKernel.cpp`. I can move that to `native/cpu/zmath.h` though. Where do you think I should have the higher level functions? Currently they are in `native/ComplexHelper.cpp`.

@albanD Sounds good, I can just follow what @mruberry said about the CUDA implementation and write tests for that then.

anjali411 commented on aten/src/ATen/native/ComplexHelper.cpp:
```diff
@@ -0,0 +1,54 @@
+              "Expected object of scalar type ", real.scalar_type(), " but got "
+              "scalar type ", imag.scalar_type(), " for argument 'imag'");
+  // Sort of hacky, but necessary so that input dtypes don't get promoted.
+  auto iter = TensorIterator::comparison_op(result, real, imag,
```
what is this for?

anjali411 replied:
ahh I see to do all the required iterator stuff

anjali411 replied:
what if `real` and `imag` were complex tensors? that shouldn't be allowed

wjgan7 replied:
I think because the kernel is only dispatched for floating types it'll say e.g. `"complex_cpu" not implemented for 'ComplexFloat'`. This is the same behavior as if I called `torch.cos(x)` where `x` was an integer tensor.

anjali411 commented on aten/src/ATen/native/ComplexHelper.cpp:
```diff
@@ -0,0 +1,54 @@
+DEFINE_DISPATCH(complex_polar_stub);
+
+Tensor& complex_out(Tensor& result, const Tensor& real, const Tensor& imag) {
+  TORCH_CHECK(real.scalar_type() == imag.scalar_type(),
```
why should this be necessary? I think we should just do type promotion here

wjgan7 replied:
removed.

anjali411 commented on aten/src/ATen/native/ComplexHelper.cpp:
```diff
@@ -0,0 +1,54 @@
+  TORCH_CHECK(real.scalar_type() == imag.scalar_type(),
+              "Expected object of scalar type ", real.scalar_type(), " but got "
+              "scalar type ", imag.scalar_type(), " for argument 'imag'");
+  // Sort of hacky, but necessary so that input dtypes don't get promoted.
```
there should be a check for shapes
```
TORCH_CHECK(real.sizes() == imag.sizes(),
      "complex expects real and imag tensors to have equal size, but got ", real.sizes() ,
      " and ", imag.sizes());
```

anjali411 replied:
and then we should do: `result.resize_(real.sizes());` which resizes the result if required

wjgan7 replied:
Hmm right now when I do `torch.complex(x, y)` and they are of different sizes, I get for example `The size of tensor a (4) must match the size of tensor b (2) at non-singleton dimension 1`. This is the exact same as if I called `torch.add(x, y)`. I'm guessing it's handled by TensorIterator, which I think handles the resizing too.

anjali411 commented on aten/src/ATen/native/ComplexHelper.cpp:
```diff
@@ -0,0 +1,54 @@
+
+Tensor complex(const Tensor& real, const Tensor& imag) {
+  c10::TensorOptions options = real.options();
+  switch (real.scalar_type()) {
```
I don't think this is correct. we should instead do:
```
switch (promoteTypes(t.scalar_type(), src.scalar_type()) {
   case c10::kDouble: options = options.dtype(c10::kComplexDouble); break;
   default:
      options.dtype(c10::kComplexFloat); break;
};

auto result = at::empty(real.sizes(), options);
```

wjgan7 replied:
sounds good!

anjali411 commented on aten/src/ATen/native/ComplexHelper.cpp:
```diff
@@ -0,0 +1,54 @@
+}
+
+Tensor& complex_polar_out(Tensor& result, const Tensor& abs, const Tensor& angle) {
+  TORCH_CHECK(abs.scalar_type() == angle.scalar_type(),
```
same comments as above

anjali411 commented:
@wjgan7 any updates?

wjgan7 committed:
Merge branch 'master' of github.com:pytorch/pytorch into polar

For some reason old branch isn't compiling properly anymore

wjgan7 commented:
@anjali411 Apologies for delays, will try to get an update soon.

wjgan7 committed:
Small changes to complex/complex_polar and CUDA

wjgan7 committed:
Fixing lint errors

wjgan7 committed:
Updated docs

wjgan7 committed:
Cleaning up tests

wjgan7 committed:
Fix CUDA type promotion issue

wjgan7 commented:
@anjali411 @mruberry Hi guys, I added the CUDA functionality based on what @mruberry linked and made changes to the type promotion based on @anjali411 suggestions.

anjali411 commented on aten/src/ATen/native/ComplexHelper.cpp:
```diff
@@ -0,0 +1,83 @@
+  return result;
+}
+
+Tensor complex(const Tensor& real, const Tensor& imag) {
```
nit - I think these functions should be added in `aten/src/ATen/native/TensorFactories.cpp` along with other factory functions.

anjali411 commented on aten/src/ATen/native/cpu/ComplexHelperKernel.cpp:
```diff
@@ -0,0 +1,32 @@
+namespace {
+
+void complex_kernel(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_TYPES(iter.input_dtype(), "complex_cpu", [&]() {
```
We should only dispatch this for `torch.float` and `torch.double` dtypes.

wjgan7 replied:
I think AT_DISPATCH_FLOATING_TYPES only does those two (it doesn't include torch.half).

anjali411 commented on aten/src/ATen/native/cpu/ComplexHelperKernel.cpp:
```diff
@@ -0,0 +1,32 @@
+}
+
+void complex_polar_kernel(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_TYPES(iter.input_dtype(), "complex_polar_cpu", [&]() {
```
same comment as above

anjali411 commented:
We should also add entries to `derivatives.yaml` for the gradients using the formulas mentioned here https://github.com/pytorch/pytorch/pull/41012:

```
- name: complex(Tensor real, Tensor imag) -> Tensor
   real: at::real(grad)
   imag: at::imag(grad) * -1

- name: complex_polar(Tensor abs, Tensor angle) -> Tensor
   abs: at::real(grad) * angle.cos()  - at::imag(grad) * angle.sin()
   angle: (at::real(grad) * angle.cos()  + at::imag(grad) * angle.sin()) * abs * -1

```

I think we should add a kernel for backward for `complex_polar` because trigonometric functions are more expensive.

cc. @albanD

anjali411 commented:
```
Running test_overrides ... [2020-07-10 06:57:46.777896]
Jul 10 06:57:47 Traceback (most recent call last):
Jul 10 06:57:47   File "test_overrides.py", line 307, in <module>
Jul 10 06:57:47     generate_tensor_like_torch_implementations()
Jul 10 06:57:47   File "test_overrides.py", line 302, in generate_tensor_like_torch_implementations
Jul 10 06:57:47     assert len(untested_funcs) == 0, msg.format(pprint.pformat(untested_funcs))
Jul 10 06:57:47 AssertionError: The following functions are not tested for __torch_function__ support, please ensure there is an entry in the dict returned by torch._overrides.get_testing_overrides for this function or if a __torch_function__ override does not make sense, add an entry to the tuple returned by torch._overrides.get_ignored_functions.
Jul 10 06:57:47 
Jul 10 06:57:47 ["<module 'torch' from "
Jul 10 06:57:47  "'/opt/conda/lib/python3.6/site-packages/torch/__init__.py'>.complex",
Jul 10 06:57:47  "<module 'torch' from "
Jul 10 06:57:47  "'/opt/conda/lib/python3.6/site-packages/torch/__init__.py'>.complex_polar",
Jul 10 06:57:47  "<module 'torch' from "
Jul 10 06:57:47  "'/opt/conda/lib/python3.6/site-packages/torch/__init__.py'>.complex",
Jul 10 06:57:47  "<module 'torch' from "
Jul 10 06:57:47  "'/opt/conda/lib/python3.6/site-packages/torch/__init__.py'>.complex_polar"]
Jul 10 06:57:47 Traceback (most recent call last):
Jul 10 06:57:47   File "test/run_test.py", line 728, in <module>
Jul 10 06:57:47     main()
Jul 10 06:57:47   File "test/run_test.py", line 721, in main
Jul 10 06:57:47     raise RuntimeError(message)
Jul 10 06:57:47 RuntimeError: test_overrides failed!
```
you will need to add entries for `complex` and `complex_polar` [here](https://github.com/pytorch/pytorch/blob/db38487eceb6abd0f4acd8e53cfab94360d106e5/torch/_overrides.py#L149) 
to fix the test errors.

wjgan7 committed:
Moving to TensorFactories, adding overrides, and adding derivatives (with tests)

wjgan7 commented:
Hi @anjali411, made some updates. I wrote some new test cases for the derivatives.

mruberry commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -98,6 +101,46 @@ Tensor _dim_arange(const Tensor& like, int64_t dim) {
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~ complex / complex_polar ~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Tensor& complex_out(Tensor& result, const Tensor& real, const Tensor& imag) {
+  // Sort of hacky, but necessary so that input dtypes don't get promoted to complex.
```
Making these a comparison op is VERY hacky and should not be done.

Just construct the appropriate TensorIteratorConfig yourself. Also, type promotion in TensorIterator now only considers input, not output, tensors, so having a complex output won't promote the input tensors to complex internally.

mruberry replied:
You might want to just verify that real and image have the same dtype, that dtype is float32 or float64, and that out has the correct corresponding complex type here.

wjgan7 replied:
Sounds good. About the dtypes, I originally did that but @anjali411 suggested I have type promotion. Should I switch back? I'll add the out check.

mruberry replied:
Adding type promotion wouldn't be hard but I'd suggest you simplify your life and we worry about that later (and separately) if it comes up.

wjgan7 replied:
It's already implemented in this PR.

wjgan7 replied:
Oh I think I see what you mean, I don't have it implemented for `complex_out`. It seems the use of a separate `complex_out` here doesn't add much value while being tedious for checks. Does every PyTorch function need to support the out argument? If not, would an exception be suitable here?

mruberry replied:
It's typical to support `out` although you're right that the value added is often small (`out` is not popular). I think it's OK to not support an `out` argument to start, although I don't know how much easier that makes things.

wjgan7 replied:
I see. What I'm thinking right now is that `complex` just uses `real`s dtype and the function for converting from float to complex (maybe the function can throw an error if `real`'s dtype isn't floating) to construct `result`'s dtype. Then `complex_out` will check same dtype, that `real` is floating, and then that `result` is the corresponding complex dtype via the same function from earlier. Any thoughts?

mruberry replied:
That sounds correct.

wjgan7 replied:
Cool. I'll try to find if the function's already implemented.

mruberry replied:
It is:

https://github.com/pytorch/pytorch/blob/e6221f4ca16e7d9fa9179db625d8dea7ea6af02b/c10/core/DefaultDtype.cpp#L20

mruberry replied:
Oh wait, you need a slightly different function, sorry. @anjali411 can tell you if there's a function mapping from float type to corresponding complex type.

wjgan7 replied:
I'm a bit confused, should I be calling `set_default_dtype(real.scalar_type())` and then `get_default_complex_dtype()`?

mruberry replied:
No no no: absolutely not. Sorry, like I said, I was mistaken about the function you need. You do want a function mapping the given float dtypes of the input tensors to the corresponding complex type.

wjgan7 replied:
Ah didn't see your second comment, got it.

mruberry commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -98,6 +101,46 @@ Tensor _dim_arange(const Tensor& like, int64_t dim) {
+
+Tensor complex(const Tensor& real, const Tensor& imag) {
+  c10::TensorOptions options = real.options();
+  switch (promote_types(real.scalar_type(), imag.scalar_type())) {
```
I think @anjali411 wrote a function that maps float types to their corresponding complex types.

Should this throw a runtime error if given anything other than float32 or float64 tensors? You can just require real and imag have the same dtype, too.

mruberry commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -98,6 +101,46 @@ Tensor _dim_arange(const Tensor& like, int64_t dim) {
+}
+
+Tensor& complex_polar_out(Tensor& result, const Tensor& abs, const Tensor& angle) {
+  // Sort of hacky, but necessary so that input dtypes don't get promoted to complex.
```
See above comment.

mruberry commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -98,6 +101,46 @@ Tensor _dim_arange(const Tensor& like, int64_t dim) {
+
+Tensor complex_polar(const Tensor& abs, const Tensor& angle) {
+  c10::TensorOptions options = abs.options();
+  switch (promote_types(abs.scalar_type(), angle.scalar_type())) {
```
See above comment.

mruberry commented on aten/src/ATen/native/cpu/ComplexKernel.cpp:
```diff
@@ -0,0 +1,32 @@
+REGISTER_DISPATCH(complex_stub, &complex_kernel);
+REGISTER_DISPATCH(complex_polar_stub, &complex_polar_kernel);
+
+}} // namespace at::native
```
newline

wjgan7 committed:
Different dtype behavior, more checks, updated tests, updated docs

wjgan7 committed:
clang-format

wjgan7 commented:
@mruberry Things I'm a bit unsure about

- `toComplexType`'s error message isn't that nice. Is there a way we can catch the TORCH_CHECK and reraise a different error?
- in test/test_complex.py, I define this function to map strings like torch.complex128 to strings like ComplexDouble. Not sure if that's ugly or not.

mruberry commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -98,6 +101,80 @@ Tensor _dim_arange(const Tensor& like, int64_t dim) {
+      .add_output(out)
+      .add_input(a)
+      .add_input(b)
+      .allow_cpu_scalars(true)
```
Does this really support calls like `torch.complex(t, 5)`?

wjgan7 replied:
hmm oh it doesn't actually.

mruberry replied:
That's OK. You can just remove this line, I think.

mruberry commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -98,6 +101,80 @@ Tensor _dim_arange(const Tensor& like, int64_t dim) {
+}
+
+Tensor complex(const Tensor& real, const Tensor& imag) {
+  c10::TensorOptions options = real.options();
```
Per your comments, if you'd like a nicer error message you can TORCH_CHECK the dtypes of the inputs here and in `complex_out` above manually. In particular, you could:

- verify the dtypes of real and imag are the same
- verify the dtype of real is torch.float32 or torch.float64, and if not say that complex doesn't support inputs the a dtype of X (whatever real's dtype is)

I think that would look nicer?

wjgan7 replied:
Yes but my concern is that the second bullet point "seems" more important. If real is torch.int32 and imag is torch.int64, I think it's better to say "Inputs are not floating type" as opposed to  "Expected object of scalar type torch.int32 but got scalar type torch.int64 for argument imag".

mruberry commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -98,6 +101,80 @@ Tensor _dim_arange(const Tensor& like, int64_t dim) {
+    Tensor& result,
+    const Tensor& abs,
+    const Tensor& angle) {
+  TORCH_CHECK(
```
You could add the same checks here. Since you'll be doing the same checks so often you could even write a helper function for them.

wjgan7 replied:
Sounds good.

mruberry commented on aten/src/ATen/native/cpu/ComplexKernel.cpp:
```diff
@@ -0,0 +1,32 @@
+void complex_polar_kernel(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_TYPES(iter.input_dtype(), "complex_polar_cpu", [&]() {
+    cpu_kernel(iter, [=](scalar_t a, scalar_t b) -> c10::complex<scalar_t> {
+      return c10::complex<scalar_t>(a * std::cos(b), a * std::sin(b));
```
I am intellectually curious if there is a faster way to compute this since we want both the cosine and sine of `b`, but I don't think it's interesting to try and compute sine as `sqrt(1 - (std::cos(b))^2`

wjgan7 replied:
There will have to be checks for sign though if sqrt is used right. I changed the kernel to this (has sign errors)
```c++
void complex_polar_kernel(TensorIterator& iter) {
  AT_DISPATCH_FLOATING_TYPES(iter.input_dtype(), "complex_polar_cpu", [&]() {
    cpu_kernel(iter, [=](scalar_t a, scalar_t b) -> c10::complex<scalar_t> {
      scalar_t acos = a * std::cos(b);
      scalar_t asin = std::sqrt(a * a - acos * acos);
      return c10::complex<scalar_t>(acos, asin);
    });
  });
}
```
and I ran this
```python
import time
import torch

a = torch.randn(100, 100)
b = torch.randn(100, 100)

start = time.time()
for i in range(50000):
    z = torch.complex_polar(a, b)
end = time.time()
print(end - start)
```
and the difference was 3.96 (current) vs 3.69 (sqrt(1-cos^2)) seconds on my computer.

mruberry replied:
Thanks for investigating - that's cool! And yes, there will be sign issues because the solution to the problem is ambiguous. Interesting that it would seem to make a significant performance improvement, though, if the sign issue could be resolved. I don't think we need to be so performance sensitive with this PR, however.

anjali411 replied:
Interesting I think we can do something like this:
```
void complex_polar_kernel(TensorIterator& iter) {
  AT_DISPATCH_FLOATING_TYPES(iter.input_dtype(), "complex_polar_cpu", [&]() {
    cpu_kernel(iter, [=](scalar_t a, scalar_t b) -> c10::complex<scalar_t> {
      scalar_t cos_val = std::cos(b);
      scalar_t sin_val = std::sqrt(1 - cos_val * cos_val);
      if (b < 0) {
         sin_val = sin_val * -1;
      }
      return c10::complex<scalar_t>(a*cos_val, a*sin_val);
    });
  });
}
```

@wjgan7 can you check if there's a performance boost for this?

mruberry commented on aten/src/ATen/native/cuda/ComplexKernel.cu:
```diff
@@ -0,0 +1,49 @@
+namespace {
+
+void complex_kernel_cuda(TensorIterator& iter) {
+  // Necessary since comparison_op doesn't actually promote types for CUDA
```
This comment is outdated, right?

wjgan7 replied:
Yep, thanks.

mruberry commented on aten/src/ATen/native/cuda/ComplexKernel.cu:
```diff
@@ -0,0 +1,49 @@
+  // Necessary since comparison_op doesn't actually promote types for CUDA
+  // tensors
+  AT_DISPATCH_FLOATING_TYPES(
+      promote_types(iter.input_dtype(0), iter.input_dtype(1)),
```
You don't want to promote the types here, right? In the earlier functions you're already guaranteeing these have the same dtype, either float32 or float64.

wjgan7 replied:
Yep, thanks.

mruberry commented on aten/src/ATen/native/cuda/ComplexKernel.cu:
```diff
@@ -0,0 +1,49 @@
+  // Necessary since comparison_op doesn't actually promote types for CUDA
+  // tensors
+  AT_DISPATCH_FLOATING_TYPES(
+      promote_types(iter.input_dtype(0), iter.input_dtype(1)),
```
Same questions about promotion here.

wjgan7 replied:
Yep.

mruberry commented on aten/src/ATen/native/cuda/ComplexKernel.cu:
```diff
@@ -0,0 +1,49 @@
+      promote_types(iter.input_dtype(0), iter.input_dtype(1)),
+      "complex_cuda",
+      [&]() {
+        gpu_kernel_with_scalars(
```
Why the scalar variant?

mruberry commented on test/test_complex.py:
```diff
@@ -1,10 +1,24 @@
+import numpy as np
```
You can't unconditionally import NumPy. What you can do is import TEST_NUMPY from common_utils.py (see how test_torch.py does this) and then import numpy if TEST_NUMPY is True

mruberry replied:
(Although since you're only using NumPy for pi below, as best I can tell, you probably just want to switch to using math.pi.)

mruberry commented on test/test_complex.py:
```diff
@@ -21,6 +35,111 @@ def test_dtype_inference(self, device, dtype):
+        complex_dtype = torch.complex64 if dtype == torch.float32 else torch.complex128
+        self.assertEqual(torch.tensor([1.0 + 3.0j, 2.0 + 4.0j], dtype=complex_dtype), z)
+
+    @dtypes(torch.float32, torch.float64)
```
This test needs a decorator to be skipped if TEST_NUMPY is False. Again, check out test_torch.py where they have decorators to do this.

mruberry commented on test/test_complex.py:
```diff
@@ -21,6 +35,111 @@ def test_dtype_inference(self, device, dtype):
+    @dtypes(torch.float32, torch.float64)
+    def test_torch_complex_polar(self, device, dtype):
+        abs = torch.tensor([1, 2], device=device, dtype=dtype)
+        angle = torch.tensor([np.pi / 2, 5 * np.pi / 4], device=device, dtype=dtype)
```
Although instead of using numpy here I think you can just use math.pi?

mruberry replied:
Better add a few more angles, too: 0, pi/4, pi/2, pi, -pi/4, -pi/2, -pi.

wjgan7 replied:
Added a few more but slightly different examples.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1465,6 +1465,47 @@ def merge_dicts(*dicts):
 
+add_docstr(torch.complex,
+           r"""
+complex(real, imag, out=None) -> Tensor
```
real, imag, *, out=None

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1465,6 +1465,47 @@ def merge_dicts(*dicts):
+           r"""
+complex(real, imag, out=None) -> Tensor
+
+Constructs a complex-valued tensor from Cartesian coordinates.
```
Constructs a complex tensor with its real part equal to :attr:`real` and its imaginary part equal to :attr:`imag`.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1465,6 +1465,47 @@ def merge_dicts(*dicts):
+:attr:`real` and :attr:`imag` should be floating-point tensors of the same
+dtype. ``torch.float32`` tensors will lead to an output of ``torch.complex64``,
+and ``torch.float64`` tensors will lead to an output of ``torch.complex128``.
+
```
Add "Args:" and "Keyword args:" (for out) here. See what `torch.add` does as a reference.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1465,6 +1465,47 @@ def merge_dicts(*dicts):
+
+add_docstr(torch.complex_polar,
+           r"""
+complex_polar(abs, angle, out=None) -> Tensor
```
abs, angle, *, out=None

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1465,6 +1465,47 @@ def merge_dicts(*dicts):
+           r"""
+complex_polar(abs, angle, out=None) -> Tensor
+
+Constructs a complex-valued tensor from polar coordinates.
```
Maybe something like:

Constructs a complex tensor whose elements are Cartesian coordinates corresponding to the polar coordinates with magnitude magnitude :attr:`abs` and angle :attr:`angle`. 

Follow-up question: should "abs" be "abs" or something like "magnitude?" cc @anjali411 to review naming. 

You could even show the math for this one, if you want to write some tex.

wjgan7 replied:
Sure. About the naming, I think abs/angle is what's used elsewhere in the code right now (I see it in zmath.h and vec256_base.h). An alternative is mag/phase, which I guess removes the conflict with Python's built in abs.

mruberry replied:
In those other contexts I think "abs" stands for absolute value, which is not here, it's just that usually in polar coordinates we don't think of the "distance" or "magnitude" term as being signed since it plus the angle are a vector.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1465,6 +1465,47 @@ def merge_dicts(*dicts):
+:attr:`abs` and :attr:`angle` should be floating-point tensors of the same
+dtype. ``torch.float32`` tensors will lead to an output of ``torch.complex64``,
+and ``torch.float64`` tensors will lead to an output of ``torch.complex128``.
+
```
Like the previous doc entry, this needs "Args" and "Keyword args"

mruberry reviewed (changes requested):
Hey @wjgan7, thanks for the updates!

I still have some questions about how this is doing, or not doing, type promotion, and whether it can be simplified. I tried to answer your question about improving the error message, too. Thank you for caring so much about the user's experience. 

As for the dtype name function, one way to work around that issue would be to simply stop checking the strings. Short of that I think you need it. In the future the test suite should probably provide that mechnaism for you so you don't need to write your own.

I also added some comments + questions to the tests and docs. Let me know what you think.

wjgan7 committed:
Addressing comments

wjgan7 commented:
@mruberry thanks for the comments, most of them should be addressed now. About the dtype checking, I ended up adding my own helper function with a different message for non floating inputs. There is a bit of redundancy but overall I think it's good.

mruberry commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -98,6 +101,75 @@ Tensor _dim_arange(const Tensor& like, int64_t dim) {
+              " for argument 'out'");
+}
+
+TensorIterator complex_op(
```
Better inline this function - it's not saving any lines

wjgan7 committed:
Inline + better formatting in .cu

mruberry commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -98,6 +101,75 @@ Tensor _dim_arange(const Tensor& like, int64_t dim) {
+              " for argument 'out'");
+}
+
+inline TensorIterator complex_op(
```
Sorry, I was called away during my last review. What I meant by "inlining" this function is to just eliminate the function and duplicate the code constructing the tensor iterator.

mruberry commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -98,6 +101,75 @@ Tensor _dim_arange(const Tensor& like, int64_t dim) {
+void complex_check_floating(const Tensor& a, const Tensor& b) {
+  TORCH_CHECK((a.scalar_type() == kFloat || a.scalar_type() == kDouble) &&
+              (b.scalar_type() == kFloat || b.scalar_type() == kDouble),
+              "Both inputs must be either Float or Double");
```
"Expected both inputs to be float or double tensors, but got ..."

Similar check for reference: 

https://github.com/pytorch/pytorch/blob/ba98c0e38ca76a8e03d27e20a1d2bfc324d502ca/aten/src/ATen/native/cuda/Lerp.cu#L11

mruberry commented on test/test_complex.py:
```diff
@@ -21,6 +35,113 @@ def test_dtype_inference(self, device, dtype):
+    def test_torch_complex_floating_dtype_error(self, device, dtype):
+        real = torch.tensor([1, 2], device=device, dtype=dtype)
+        imag = torch.tensor([3, 4], device=device, dtype=dtype)
+        error = "Both inputs must be either Float or Double"
```
This check will need to be updated

mruberry commented on test/test_complex.py:
```diff
@@ -21,6 +35,113 @@ def test_dtype_inference(self, device, dtype):
+
+    @dtypes(torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64,
+            torch.float16, torch.complex64, torch.complex128, torch.bool)
+    def test_torch_complex_floating_dtype_error(self, device, dtype):
```
You can combine this test and the next one by starting with a for loop:

```
for op in (torch.complex, torch.complex_polar):
```

mruberry commented on test/test_complex.py:
```diff
@@ -21,6 +35,113 @@ def test_dtype_inference(self, device, dtype):
+        with self.assertRaisesRegex(RuntimeError, error):
+            torch.complex_polar(abs, angle)
+
+    @dtypes(torch.float32, torch.float64)
```
This test and the next one can also be combined by iterating over the ops

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1465,6 +1465,69 @@ def merge_dicts(*dicts):
+        :attr:`real`.
+
+Keyword arguments:
+    {out} If the inputs are ``torch.float32``, must be ``torch.complex64``. If
```
"out (Tensor):"  like the above arguments

wjgan7 replied:
I believe it's substituted by common_args

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1465,6 +1465,69 @@ def merge_dicts(*dicts):
+Args:
+    real (Tensor): The real part of the complex tensor. Must be
+        ``torch.float32`` or ``torch.float64``.
+    imag (Tensor): The imaginary part of the complex tensor. Must be
```
"Must be the same dtype as :attr:`real`" is enough, since you've already specified that real must be float32 or float64.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1465,6 +1465,69 @@ def merge_dicts(*dicts):
+    \text{out} = \text{real} + \text{imag} \cdot i
+""" + r"""
+Args:
+    real (Tensor): The real part of the complex tensor. Must be
```
"Must be a float or double tensor." might be a more natural sounding second sentence.

wjgan7 replied:
After reading it a few more times I think "Must be float or double." is even better.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1465,6 +1465,69 @@ def merge_dicts(*dicts):
+imaginary part equal to :attr:`imag`.
+
+.. math::
+    \text{out} = \text{real} + \text{imag} \cdot i
```
Our complex numbers are a + bj, not a + bi.

mruberry replied:
That said, I don't know if you need a math example here. The center dot notation may not be super clear, anyway.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1465,6 +1465,69 @@ def merge_dicts(*dicts):
+        or ``torch.float64`` and furthermore of same dtype as :attr:`abs`.
+
+Keyword arguments:
+    {out} If the inputs are ``torch.float32``, must be ``torch.complex64``. If
```
See comments above.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1465,6 +1465,69 @@ def merge_dicts(*dicts):
+    angle (Tensor): The angle of the complex tensor. Must be ``torch.float32``
+        or ``torch.float64`` and furthermore of same dtype as :attr:`abs`.
+
+Keyword arguments:
```
Nit: "Keyword args:"

mruberry commented:
Nice updates, @wjgan7! I made a few minor suggestions, mostly for the test and docs.

@anjali411, would you like to take a look? I know you've been thinking about the use of trig functions in complex kernels. The doc presentation and complex backward could also use your review.

What do you think of "abs" vs. "magnitude" or "distance" for the argument names for complex_polar?

anjali411 commented on aten/src/ATen/native/cpu/ComplexKernel.cpp:
```diff
@@ -0,0 +1,32 @@
+  AT_DISPATCH_FLOATING_TYPES(iter.input_dtype(), "complex_cpu", [&]() {
+    cpu_kernel(iter, [=](scalar_t a, scalar_t b) -> c10::complex<scalar_t> {
+      return c10::complex<scalar_t>(a, b);
+    });
```
cc. @VitalyFedyunin do you have an idea of how we can add vectorization here? 
```
[=](Vec256<scalar_t> a, Vec256<scalar_t> a) -> std::pair<Vec256<c10::complex<scalar_t>>, Vec256<c10::complex<scalar_t>>> { ... });
```
this wouldn't work with the current design of `cpu_kernel` right?

VitalyFedyunin replied:
It should be `[=](Vec256<scalar_t> a, Vec256<scalar_t> a) -> Vec256<c10::complex<scalar_t>> { ... };` instead

VitalyFedyunin replied:
Sorry I take it back, you will need custom loop which is to complex to use here.

mruberry replied:
Could you use a kernel to interleave the float values, then reinterpret the storage as complex? 

That is you have inputs a0 a1 a2 and b0 b1 b2, the kernel interleaves to a0 b0 a1 b1 a2 b2, and then outside the kernel the storage is reinterpreted as complex.

anjali411 replied:
yeah I like that idea however we would still need a custom kernel since we will still have two `Vec256<c10::complex<scalar_t>>` vectors as output for two `Vec256<scalar_t>` vector input.

anjali411 replied:
@wjgan7 just to be clear, adding vectorization for `torch.complex` and `torch.polar` is out of the scope of this PR and can be added in a follow up PR.

mruberry commented:
Maybe interesting to look at C++'s naming for similar functions: https://en.cppreference.com/w/cpp/numeric/complex/polar

wjgan7 committed:
Inlining, new float error msg, combining error tests

wjgan7 committed:
Simplifying docs language, removing some math, nits

wjgan7 commented:
@mruberry Thanks for the additional feedback, addressed most of the comments. I'm open to any naming scheme for `complex_polar`

anjali411 commented:
> Maybe interesting to look at C++'s naming for similar functions: https://en.cppreference.com/w/cpp/numeric/complex/polar

that's a good point! [cmath](https://docs.python.org/2/library/cmath.html) also has `torch.polar` so I think it's best to go with `torch.polar` so that the naming is consistent with the standard C++ and python libraries.

anjali411 commented on aten/src/ATen/core/aten_interned_strings.h:
```diff
@@ -246,6 +246,8 @@ _(aten, clamp_min) \
 _(aten, coalesce) \
 _(aten, combinations) \
+_(aten, complex) \
+_(aten, complex_polar) \
```
let's change this to `polar` (following the C++ std and cmath naming convention)

anjali411 commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -98,6 +101,74 @@ Tensor _dim_arange(const Tensor& like, int64_t dim) {
   return at::arange(like.size(dim), like.options().dtype(at::kLong));
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~ complex / complex_polar ~~~~~~~~~~~~~~~~~~~~~~~~~~
```
not complex / polar

anjali411 commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -64,6 +64,9 @@ static inline bool allIntegral(std::initializer_list<std::reference_wrapper<Scal
 } // namespace
 
+DEFINE_DISPATCH(complex_stub);
+DEFINE_DISPATCH(complex_polar_stub);
```
`DEFINE_DISPATCH(polar_stub);`

anjali411 commented on test/test_complex.py:
```diff
@@ -21,6 +35,84 @@ def test_dtype_inference(self, device, dtype):
         torch.set_default_dtype(default_dtype)
         self.assertEqual(x.dtype, torch.cdouble if dtype == torch.float64 else torch.cfloat)
 
+    @dtypes(torch.float32, torch.float64)
```
All these tests should go in `test_torch.py`

anjali411 commented on test/test_complex.py:
```diff
@@ -1,10 +1,24 @@
 devices = (torch.device('cpu'), torch.device('cuda:0'))
 
 
+def dtype_name(dtype):
```
no updates should be made in `test_complex.py`

anjali411 commented on torch/_torch_docs.py:
```diff
@@ -1465,6 +1465,64 @@ def merge_dicts(*dicts):
+
+Constructs a complex tensor whose elements are Cartesian coordinates
+corresponding to the polar coordinates with absolute value / magnitude
+:attr:`abs` and angle :attr:`angle`.
```
nit - angle and abs are functions :func:`abs` and :func:`angle`

anjali411 commented on torch/_torch_docs.py:
```diff
@@ -1465,6 +1465,64 @@ def merge_dicts(*dicts):
+complex_polar(abs, angle, *, out=None) -> Tensor
+
+Constructs a complex tensor whose elements are Cartesian coordinates
+corresponding to the polar coordinates with absolute value / magnitude
```
nit - just absolute value is fine and we can get rid of the magnitude since absolute value for complex numbers has a non ambiguous definition.

anjali411 commented:
Left some minor comments. This PR is on the right trajectory and should be ready to go once the comments (naming, testing an alternate way for perf gains, and moving tests to `test_torch.py`) are addressed.

wjgan7 committed:
complex_polar -> polar, move tests to test_torch.py

wjgan7 commented:
@anjali411 Addressed the comments. About the performance thing, I got 3.967 s on the original vs 5.895 s with the new code, running the same test as before.

anjali411 commented:
@wjgan7 thanks! LGTM. Once the merge conflicts are resolved and CI is green, it'll be good to go.

wjgan7 committed:
Fix PI import issue after moving to test_torch.py

wjgan7 committed:
Temporarily removing docs in hopes of being able to merge

wjgan7 committed:
Merge branch 'master' into polar

wjgan7 committed:
Adding back in docs

anjali411 commented on test/test_torch.py:
```diff
@@ -7114,6 +7114,101 @@ def test_clamp(self, device):
+
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.float32, torch.float64)
+    def test_torch_complex_backward(self, device, dtype):
```
I am sorry I didn't look at this before , but the autograd tests shouldn't exist in `test_torch.py`. Since, we don't have `gradcheck` for complex yet, these tests can be added in `TestAutogradComplex` in `test_autograd.py`. 
(self note: should be modified later to use gradcheck)

wjgan7 replied:
I see, seems like `TestAutogradComplex` doesn't support device/dtype, so I removed those from the tests. I guess this is fine since the derivative is based on that one line we put in `derivatives.yaml`, so it shouldn't depend on device/dtype anyway. I also added `# remove this test after gradcheck support is added for non-holomorphic functions` comments since other tests use them and I'm guessing that's what you mean by the gradcheck stuff.

anjali411 reviewed (changes requested):
LGTM after the autograd tests have been moved from `test_torch.py` to `test_autograd.py`.

cc. @mruberry to take a final look!

anjali411 commented on test/test_torch.py:
```diff
@@ -7114,6 +7114,101 @@ def test_clamp(self, device):
+        z = torch.complex(real, imag)
+        loss = z.sum()
+        loss.backward()
+        print(real.grad)
```
nit -- remove `print(real.grad)`

wjgan7 replied:
Thanks for catching!

wjgan7 committed:
Moving autograd tests to test_autograd

wjgan7 committed:
Removing decorators for autograd tests

anjali411 commented:
Synced with @albanD offline. We shouldn't add any new autograd formulas until [this](https://github.com/pytorch/pytorch/issues/41857) issue is resolved so we should just add `not_implemented("fn_name")` in `derivatives.yaml` definitions.

anjali411 commented on tools/autograd/derivatives.yaml:
```diff
@@ -329,6 +329,14 @@
   self: grad
 
+- name: complex(Tensor real, Tensor imag) -> Tensor
+  real: at::real(grad)
```
'not_implemented("complex: real")'

anjali411 replied:
@wjgan7 I don't think this has been resolved. you need to replace 
`real: at::real(grad)` with 'real: not_implemented("complex: real")'

wjgan7 replied:
Oops should be there now.

anjali411 commented on tools/autograd/derivatives.yaml:
```diff
@@ -329,6 +329,14 @@
 
+- name: complex(Tensor real, Tensor imag) -> Tensor
+  real: at::real(grad)
+  imag: at::imag(grad) * -1
```
'not_implemented("complex: imag")'

anjali411 replied:
same as above

anjali411 commented on tools/autograd/derivatives.yaml:
```diff
@@ -329,6 +329,14 @@
+  imag: at::imag(grad) * -1
+
+- name: polar(Tensor abs, Tensor angle) -> Tensor
+  abs: at::real(grad) * angle.cos()  - at::imag(grad) * angle.sin()
```
'not_implemented("polar: abs")'

anjali411 replied:
same as above

anjali411 commented on tools/autograd/derivatives.yaml:
```diff
@@ -329,6 +329,14 @@
+
+- name: polar(Tensor abs, Tensor angle) -> Tensor
+  abs: at::real(grad) * angle.cos()  - at::imag(grad) * angle.sin()
+  angle: (at::real(grad) * angle.cos()  + at::imag(grad) * angle.sin()) * abs * -1
```
'not_implemented("polar: angle")'

anjali411 replied:
same as above

anjali411 commented on test/test_complex.py:
```diff
@@ -21,6 +21,7 @@ def test_dtype_inference(self, device, dtype):
         torch.set_default_dtype(default_dtype)
         self.assertEqual(x.dtype, torch.cdouble if dtype == torch.float64 else torch.cfloat)
 
+
```
nit - remove this blank line

anjali411 commented on test/test_autograd.py:
```diff
@@ -4734,6 +4734,26 @@ def func(z):
         self.assertEqual(z.grad, z1.grad)
 
+    # remove this test after gradcheck support is added for non-holomorphic functions
+    def test_torch_complex_backward(self):
```
remove both these tests since we are not adding any formulas in `derivatives.yaml`.

anjali411 reviewed (changes requested):
Synced with @albanD offline. We shouldn't add any new autograd formulas until JAX vs tf autograd issue is resolved so we should just add not_implemented("fn_name") in derivatives.yaml definitions.

wjgan7 commented:
@anjali411 Not sure if I can view the failing CircleCI tests. Do you know what the issue is?

anjali411 commented:
@wjgan7 they are just upstream failures, but if you rebase on the latest master, CI should be green.

wjgan7 committed:
Accidentally didn't add files to commit

anjali411 reviewed (approved):
LGTM :) thanks for the awesome work @wjgan7

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1502,6 +1502,63 @@ def merge_dicts(*dicts):
+        as :attr:`real`.
+
+Keyword args:
+    {out} If the inputs are ``torch.float32``, must be ``torch.complex64``. If
```
"{out} -> out (Tensor): "

You don't want to autogenerate the out doc, which is what happens when using `{out}`

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1502,6 +1502,63 @@ def merge_dicts(*dicts):
+        :attr:`abs`.
+
+Keyword args:
+    {out} If the inputs are ``torch.float32``, must be ``torch.complex64``. If
```
Same issue with "{out}" here

mruberry reviewed (approved):
Looks good but please fix the "{out}" in the two doc updates.

wjgan7 committed:
Don't use {out}

wjgan7 commented:
@anjali411 @mruberry Thanks for all the reviews, I learned a lot!
